dask/dataframe/shuffle.py

```diff
@@ -532,6 +723,78 @@ def _simple_rearrange_by_column_tasks(df, column, npartitions, ignore_index=Fals
 
 def rearrange_by_column_tasks(
     df, column, max_branch=32, npartitions=None, ignore_index=False
+):
+    max_branch = max_branch or 32
+    n = df.npartitions
+
+    stages = int(math.ceil(math.log(n) / math.log(max_branch)))
+    if stages > 1:
+        k = int(math.ceil(n ** (1 / stages)))
+    else:
+        k = n
+
+    inputs = [tuple(digit(i, j, k) for j in range(stages)) for i in range(k ** stages)]
+
+    npartitions_orig = df.npartitions
+    token = tokenize(df, stages, column, n, k)
+    for stage in range(stages):
+        stage_name = f"shuffle-{stage}-{token}"
+        stage_layer = ShuffleStage(
+            stage_name,
+            column,
+            inputs,
+            stage,
+            npartitions,
+            n,
+            k,
+            ignore_index,
+            df._name,
+            df._meta,
+        )
+        graph = HighLevelGraph.from_collections(
+            stage_name, stage_layer, dependencies=[df]
+        )
+        df = new_dd_object(graph, stage_name, df._meta, df.divisions)
+
+    if npartitions is not None and npartitions != npartitions_orig:
+        token = tokenize(df, npartitions)
+        repartition_group_token = "repartition-group-" + token
+
+        dsk = {
+            (repartition_group_token, i): (
+                shuffle_group_2,
+                k,
+                column,
+                ignore_index,
+                npartitions,
+            )
+            for i, k in enumerate(df.__dask_keys__())
+        }
+
+        repartition_get_name = "repartition-get-" + token
+
+        for p in range(npartitions):
+            dsk[(repartition_get_name, p)] = (
+                shuffle_group_get,
+                (repartition_group_token, p % npartitions_orig),
+                p,
+            )
+
+        graph2 = HighLevelGraph.from_collections(
+            repartition_get_name, dsk, dependencies=[df]
+        )
+        df2 = new_dd_object(
+            graph2, repartition_get_name, df._meta, [None] * (npartitions + 1)
+        )
+    else:
+        df2 = df
+        df2.divisions = (None,) * (npartitions_orig + 1)
+
+    return df2
+
+
+def rearrange_by_column_tasks_original(
+    df, column, max_branch=32, npartitions=None, ignore_index=False
```

Comment:
What's the plan with original implementation?  Can it be removed?

Reply:
Yup! Completely forgot that I left the original version there while implementing the new version :)